src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,411 @@
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ * ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+#include "mlio/image_reader.h"
+
+#ifdef MLIO_BUILD_IMAGE_READER
+
+#include <fmt/format.h>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/imgproc/imgproc.hpp>
+
+#include "mlio/cpu_array.h"
+#include "mlio/data_type.h"
+#include "mlio/instance.h"
+#include "mlio/instance_batch.h"
+#include "mlio/intrusive_ptr.h"
+#include "mlio/logger.h"
+#include "mlio/parallel_data_reader.h"
+#include "mlio/record_readers/blob_record_reader.h"
+#include "mlio/record_readers/recordio_record_reader.h"
+#include "mlio/schema.h"
+
+namespace mlio {
+inline namespace v1 {
+
+image_reader::image_reader(data_reader_params rdr_prm,
+                           image_reader_params image_prm)
+    : parallel_data_reader{std::move(rdr_prm)}, params_{std::move(image_prm)}
+{
+    if (params_.image_dimensions.size() != image_dimensions_size_) {
+        throw std::invalid_argument{
+            "image_dimensions is a required parameter. "
+            "Dimensions of the output image must be entered in "
+            "(channels, height, width) format"};
+    }
+    bbh_ = effective_bad_batch_handling();
+    std::copy(params_.image_dimensions.begin(),
+              params_.image_dimensions.end(),
+              img_dims_.begin());
+}
+
+image_reader::~image_reader()
+{
+    stop();
+}
+
+intrusive_ptr<record_reader>
+image_reader::make_record_reader(data_store const &ds)
+{
+    switch (params_.img_frame) {
+    case image_frame::none:
+        return make_intrusive<blob_record_reader>(ds.open_read());
+    case image_frame::recordio:
+        return make_intrusive<detail::recordio_record_reader>(ds.open_read());
+    }
+    throw std::invalid_argument{
+        "The image_frame argument does not represent a valid value."};
+}
+
+void
+image_reader::infer_schema(instance const &)
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(feature_desc_builder{"value",
+                                            data_type::uint8,
+                                            {params().batch_size,
+                                             params_.image_dimensions[1],
+                                             params_.image_dimensions[2],
+                                             params_.image_dimensions[0]}}
+                           .build());
+    schema_ = make_intrusive<schema>(std::move(descs));
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &batch) const
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    if (tsr == nullptr) {
+        return nullptr;
+    }
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(std::move(tsr));
+    return make_intrusive<example>(schema_, std::move(tensors));
```

Comment:
Instead of `schema_` use the protected `get_schema` member function of `parallel_data_reader` to access the schema.

Reply:
done